snap/squashfs2/squashfs2.go

```diff
@@ -0,0 +1,339 @@
+		return createXzBackend(optionsBlock)
+	default:
+		return nil, fmt.Errorf("squashfs: unsupported compression type %d", sb.CompressionType)
+	}
+}
+
+// createInodeReader Instantiates a new inode metadata reader with the appropriate compression support
+func createInodeReader(stream *os.File, cb CompressionBackend, sb *internal.SuperBlock) (*metaBlockReader, error) {
+	if sb.Flags&internal.SuperBlockUncompressedInodes != 0 {
+		inodeReader := metablockReaderCreate(stream, nil, sb.InodeTable, sb.RootIno)
+		if inodeReader == nil {
+			return nil, fmt.Errorf("squashfs: failed to create inode reader")
+		}
+		return inodeReader, nil
+	} else {
+		inodeReader := metablockReaderCreate(stream, cb, sb.InodeTable, sb.RootIno)
+		if inodeReader == nil {
+			return nil, fmt.Errorf("squashfs: failed to create inode reader")
+		}
+		return inodeReader, nil
+	}
+}
+
+// SquashFS layout
+// from: https://dr-emann.github.io/squashfs/
+// ---------------
+// |  superblock   |
+// |---------------|
+// |  compression  |
+// |    options    |
+// |---------------|
+// |  datablocks   |
+// |  & fragments  |
+// |---------------|
+// |  inode table  |
+// |---------------|
+// |   directory   |
+// |     table     |
+// |---------------|
+// |   fragment    |
+// |    table      |
+// |---------------|
+// |    export     |
+// |    table      |
+// |---------------|
+// |    uid/gid    |
+// |  lookup table |
+// |---------------|
+// |     xattr     |
+// |     table     |
+// |---------------|
+func Open(path string) (*SquashFileSystem, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	// Remember to close the file when we're done if any
+	// errors happens.
+	defer func() {
+		if err != nil {
+			f.Close()
+		}
+	}()
+
+	sb, err := readSuperBlock(f)
+	if err != nil {
+		return nil, err
+	}
+
+	// handle compression type
+	cb, err := createCompressionBackend(f, sb)
+	if err != nil {
+		return nil, err
+	}
+
+	// create inode reader
+	inodeReader, err := createInodeReader(f, cb, sb)
+	if err != nil {
+		return nil, err
+	}
+
+	// create directory reader
+	directoryReader := metablockReaderCreate(f, cb, sb.DirectoryTable)
+	if directoryReader == nil {
+		return nil, fmt.Errorf("squashfs: failed to create directory reader")
+	}
+
+	sfs := &SquashFileSystem{
+		stream:          f,
+		superBlock:      sb,
+		compression:     cb,
+		inodeReader:     inodeReader,
+		directoryReader: directoryReader,
+	}
+
+	// initialize root directory right away so we can start
+	// reading from it immediately.
+	err = sfs.loadRootDirectory()
+	if err != nil {
+		sfs.Close()
+		return nil, err
+	}
+
+	return sfs, nil
+}
+
+func (sfs *SquashFileSystem) directoryCreate(node *internal.InodeDir) *directory {
+	return &directory{
+		node:   node,
+		reader: sfs.directoryReader,
+		loaded: false,
+	}
+}
+
+func (sfs *SquashFileSystem) readInodeData() ([]byte, error) {
+	typeBuffer := make([]byte, 2)
+	err := sfs.inodeReader.read(typeBuffer)
```

Comment:
It's nice you bring some input here, because I was having a hard time getting the design right on this, and I don't think I ended up with anything optimal, so I really appreciate the input on my solution. It's a nice touch to read the base inode descriptor first, and then read the rest of the extended inode afterwards.